File: decode_string.rb
def decode_string(s)
    stack = []
  
    s.each_char do |ch|
      if ch != "]"
        stack << ch
      else
        subs = ''
        while stack[-1] != '['
          subs = stack.pop + subs
        end
        stack.pop
  
        int_s = ''
        while !stack.empty? && stack[-1].match?(/\d/)
          int_s = stack.pop + int_s
        end
  
        stack << subs * int_s.to_i
      end
    end
  
    stack.join('')
end
  

File: decode_string.js
/**
 * @param {string} s
 * @return {string}
 */
var decodeString = function (s) {
  const stack = [];

  for (let ch of s) {
    if (ch !== "]") {
      stack.push(ch);
    } else {
      let char = "";
      let current = stack.pop();
      while (current !== "[") {
        char = current + char;
        current = stack.pop();
      }

      let numStr = "";
      current = stack.pop();
      while (!isNaN(current)) {
        numStr = current + numStr;
        current = stack.pop();
      }
      let num = parseInt(numStr);

      let out = "";
      while (num > 0) {
        out += char;
        num--;
      }
      stack.push(out);
    }
  }

  return stack.join("");
};
